paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,518 @@
+
+    this->output_ =
+        CreateDstMemory(mul_prim_desc, ctx, output, force_fp32_output);
+
+    return inner_product_forward(mul_prim_desc, x_memory, y_memory,
+                                 *(this->output_));
+  }
+
+  inner_product_forward::primitive_desc CreateMulPrimDesc(
+      const memory::desc &x_desc, const memory::desc &y_desc,
+      const memory::desc &dst_desc, const mkldnn::primitive_attr &mul_attr) {
+    const auto &mul_desc = inner_product_forward::desc(
+        prop_kind::forward, x_desc, y_desc, dst_desc);
+
+    return inner_product_forward::primitive_desc(mul_desc, mul_attr,
+                                                 this->engine_);
+  }
+
+  memory CreateDstMemory(
+      const inner_product_forward::primitive_desc &mul_prim_desc,
+      const ExecutionContext &ctx, Tensor *output, bool force_fp32_output) {
+    auto dst_prim_desc = mul_prim_desc.dst_primitive_desc();
+    auto buffer_size = dst_prim_desc.get_size();
+
+    if (force_fp32_output) {
+      float *output_data =
+          output->mutable_data<float>(ctx.GetPlace(), buffer_size);
+      output->set_format((memory::format)dst_prim_desc.desc().data.format);
+
+      return memory(dst_prim_desc, to_void_cast<float>(output_data));
+    } else {
+      int8_t *output_data =
+          output->mutable_data<int8_t>(ctx.GetPlace(), buffer_size);
+      output->set_format((memory::format)dst_prim_desc.desc().data.format);
+
+      return memory(dst_prim_desc, to_void_cast<int8_t>(output_data));
+    }
+  }
+};
+
+static std::string GetHash(const Tensor *input_x, const Tensor *input_y,
+                           const std::string &suffix) {
+  auto dim2str = [](const DDim &operand_dims) {
+    std::string str = "";
+    for (int i = 0; i < operand_dims.size(); ++i) {
+      str += std::to_string(operand_dims[i]) + "-";
+    }
+    return str;
+  };
+
+  std::string hash = std::to_string((unsigned)input_x->format()) +
+                     std::to_string((unsigned)input_x->type()) +
+                     dim2str(input_x->dims()) +
+                     std::to_string((unsigned)input_y->format()) +
+                     std::to_string((unsigned)input_y->type()) +
+                     dim2str(input_y->dims()) + suffix;
+
+  return hash;
+}
+
+template <typename T, typename K>
+std::shared_ptr<MulPrimitiveFactory<T, K>> GetPrimitiveFactory(
+    const MKLDNNDeviceContext &dev_ctx, const ExecutionContext &ctx,
+    const Tensor *input_x, const Tensor *input_y,
+    const mkldnn::engine &mkldnn_engine) {
+  const std::string key = GetHash(input_x, input_y, ctx.op().Output("Out"));
+
+  auto prim_creator =
+      std::static_pointer_cast<MulPrimitiveFactory<T, K>>(dev_ctx.GetBlob(key));
+  if (prim_creator == nullptr) {
+    bool enable_quant =
+        std::is_same<T, int8_t>::value || std::is_same<T, uint8_t>::value;
+    if (enable_quant) {
+      prim_creator =
+          std::make_shared<QuantMulPrimitiveFactory<T, K>>(mkldnn_engine);
```

Comment:
I have noticed you have used `ctx.Attr<bool>("force_fp32_output")` in your code. Perhaps it'd be easier to check this function here, just once, and pass the information about output datatype through template parameter to the `QuantMulPrimitiveFactory` factory so that there is much less hassle with checking what the output dt should be. The code would be shorter that way as well.  That's just something to consider.

Reply:
I tried to use 3 template parameters and it does make the code shorter, but in this way, the code seems to be more difficult to understand. so I still use 2 template parameters for now.